File: tests/mv-00/dst_01.c
#include <json-model.h>
#define JSON_MODEL_VERSION "2.0a0"

static bool json_model_1(const json_t* val, Path* path, Report* rep);
propmap_t check_model_map_tab[1];
const size_t check_model_map_size = 1;

// object $.'|'.5
static bool _jm_obj_0(const json_t* val, Path* path, Report* rep)
{
    if (! json_is_object(val))
    {
        if (rep) jm_report_add_entry(rep, "not an object [$.'|'.5]", path);
        return false;
    }
    bool res;
    int64_t must_count = 0;
    const char *prop;
    json_t *pval;
    json_object_foreach((json_t *) val, prop, pval)
    {
        Path lpath_0 = (Path) { prop, 0, path, NULL };
        if (strcmp(prop, "a") == 0)
        {
            // handle must a property
            must_count += 1;
            // $.'|'.5.a
            res = json_is_integer(pval) && json_integer_value(pval) >= 0;
            if (! res)
            {
                if (rep) jm_report_add_entry(rep, "not a 0 strict int [$.'|'.5.a]", (path ? &lpath_0 : NULL));
            }
            if (! res)
            {
                if (rep) jm_report_add_entry(rep, "invalid must property value [$.'|'.5.a]", (path ? &lpath_0 : NULL));
                return false;
            }
        }
        else if (strcmp(prop, "c") == 0)
        {
            // handle must c property
            must_count += 1;
            // $.'|'.5.c
            res = json_is_integer(pval) && json_integer_value(pval) >= 1;
            if (! res)
            {
                if (rep) jm_report_add_entry(rep, "not a 1 strict int [$.'|'.5.c]", (path ? &lpath_0 : NULL));
            }
            if (! res)
            {
                if (rep) jm_report_add_entry(rep, "invalid must property value [$.'|'.5.c]", (path ? &lpath_0 : NULL));
                return false;
            }
        }
        else if (strcmp(prop, "f") == 0)
        {
            // handle must f property
            must_count += 1;
            // $.'|'.5.f
            res = json_is_string(pval);
            if (! res)
            {
                if (rep) jm_report_add_entry(rep, "unexpected string [$.'|'.5.f]", (path ? &lpath_0 : NULL));
            }
            if (! res)
            {
                if (rep) jm_report_add_entry(rep, "invalid must property value [$.'|'.5.f]", (path ? &lpath_0 : NULL));
                return false;
            }
        }
        else
        {
            if (rep) jm_report_add_entry(rep, "no other prop expected [$.'|'.5]", (path ? &lpath_0 : NULL));
            return false;
        }
    }
    if (must_count != 3)
    {
        if (rep) jm_report_add_entry(rep, "missing must prop [$.'|'.5]", path);
        return false;
    }
    return true;
}

// object $.'|'.4
static bool _jm_obj_1(const json_t* val, Path* path, Report* rep)
{
    if (! json_is_object(val))
    {
        if (rep) jm_report_add_entry(rep, "not an object [$.'|'.4]", path);
        return false;
    }
    bool res;
    int64_t must_count = 0;
    const char *prop;
    json_t *pval;
    json_object_foreach((json_t *) val, prop, pval)
    {
        Path lpath_1 = (Path) { prop, 0, path, NULL };
        if (strcmp(prop, "a") == 0)
        {
            // handle must a property
            must_count += 1;
            // $.'|'.4.a
            res = json_is_integer(pval) && json_integer_value(pval) >= 0;
            if (! res)
            {
                if (rep) jm_report_add_entry(rep, "not a 0 strict int [$.'|'.4.a]", (path ? &lpath_1 : NULL));
            }
            if (! res)
            {
                if (rep) jm_report_add_entry(rep, "invalid must property value [$.'|'.4.a]", (path ? &lpath_1 : NULL));
                return false;
            }
        }
        else if (strcmp(prop, "c") == 0)
        {
            // handle must c property
            must_count += 1;
            // $.'|'.4.c
            res = json_is_integer(pval) && json_integer_value(pval) >= 1;
            if (! res)
            {
                if (rep) jm_report_add_entry(rep, "not a 1 strict int [$.'|'.4.c]", (path ? &lpath_1 : NULL));
            }
            if (! res)
            {
                if (rep) jm_report_add_entry(rep, "invalid must property value [$.'|'.4.c]", (path ? &lpath_1 : NULL));
                return false;
            }
        }
        else if (strcmp(prop, "e") == 0)
        {
            // handle must e property
            must_count += 1;
            // $.'|'.4.e
            res = json_is_string(pval);
            if (! res)
            {
                if (rep) jm_report_add_entry(rep, "unexpected string [$.'|'.4.e]", (path ? &lpath_1 : NULL));
            }
            if (! res)
            {
                if (rep) jm_report_add_entry(rep, "invalid must property value [$.'|'.4.e]", (path ? &lpath_1 : NULL));
                return false;
            }
        }
        else
        {
            if (rep) jm_report_add_entry(rep, "no other prop expected [$.'|'.4]", (path ? &lpath_1 : NULL));
            return false;
        }
    }
    if (must_count != 3)
    {
        if (rep) jm_report_add_entry(rep, "missing must prop [$.'|'.4]", path);
        return false;
    }
    return true;
}

// object $.'|'.3
static bool _jm_obj_2(const json_t* val, Path* path, Report* rep)
{
    if (! json_is_object(val))
    {
        if (rep) jm_report_add_entry(rep, "not an object [$.'|'.3]", path);
        return false;
    }
    bool res;
    int64_t must_count = 0;
    const char *prop;
    json_t *pval;
    json_object_foreach((json_t *) val, prop, pval)
    {
        Path lpath_2 = (Path) { prop, 0, path, NULL };
        if (strcmp(prop, "a") == 0)
        {
            // handle must a property
            must_count += 1;
            // $.'|'.3.a
            res = json_is_integer(pval) && json_integer_value(pval) >= 0;
            if (! res)
            {
                if (rep) jm_report_add_entry(rep, "not a 0 strict int [$.'|'.3.a]", (path ? &lpath_2 : NULL));
            }
            if (! res)
            {
                if (rep) jm_report_add_entry(rep, "invalid must property value [$.'|'.3.a]", (path ? &lpath_2 : NULL));
                return false;
            }
        }
        else if (strcmp(prop, "c") == 0)
        {
            // handle must c property
            must_count += 1;
            // $.'|'.3.c
            res = json_is_integer(pval) && json_integer_value(pval) >= 1;
            if (! res)
            {
                if (rep) jm_report_add_entry(rep, "not a 1 strict int [$.'|'.3.c]", (path ? &lpath_2 : NULL));
            }
            if (! res)
            {
                if (rep) jm_report_add_entry(rep, "invalid must property value [$.'|'.3.c]", (path ? &lpath_2 : NULL));
                return false;
            }
        }
        else if (strcmp(prop, "d") == 0)
        {
            // handle must d property
            must_count += 1;
            // $.'|'.3.d
            res = json_is_string(pval);
            if (! res)
            {
                if (rep) jm_report_add_entry(rep, "unexpected string [$.'|'.3.d]", (path ? &lpath_2 : NULL));
            }
            if (! res)
            {
                if (rep) jm_report_add_entry(rep, "invalid must property value [$.'|'.3.d]", (path ? &lpath_2 : NULL));
                return false;
            }
        }
        else
        {
            if (rep) jm_report_add_entry(rep, "no other prop expected [$.'|'.3]", (path ? &lpath_2 : NULL));
            return false;
        }
    }
    if (must_count != 3)
    {
        if (rep) jm_report_add_entry(rep, "missing must prop [$.'|'.3]", path);
        return false;
    }
    return true;
}

// object $.'|'.2
static bool _jm_obj_3(const json_t* val, Path* path, Report* rep)
{
    if (! json_is_object(val))
    {
        if (rep) jm_report_add_entry(rep, "not an object [$.'|'.2]", path);
        return false;
    }
    bool res;
    int64_t must_count = 0;
    const char *prop;
    json_t *pval;
    json_object_foreach((json_t *) val, prop, pval)
    {
        Path lpath_3 = (Path) { prop, 0, path, NULL };
        if (strcmp(prop, "a") == 0)
        {
            // handle must a property
            must_count += 1;
            // $.'|'.2.a
            res = json_is_integer(pval) && json_integer_value(pval) >= 0;
            if (! res)
            {
                if (rep) jm_report_add_entry(rep, "not a 0 strict int [$.'|'.2.a]", (path ? &lpath_3 : NULL));
            }
            if (! res)
            {
                if (rep) jm_report_add_entry(rep, "invalid must property value [$.'|'.2.a]", (path ? &lpath_3 : NULL));
                return false;
            }
        }
        else if (strcmp(prop, "b") == 0)
        {
            // handle must b property
            must_count += 1;
            // $.'|'.2.b
            res = json_is_integer(pval) && json_integer_value(pval) >= 1;
            if (! res)
            {
                if (rep) jm_report_add_entry(rep, "not a 1 strict int [$.'|'.2.b]", (path ? &lpath_3 : NULL));
            }
            if (! res)
            {
                if (rep) jm_report_add_entry(rep, "invalid must property value [$.'|'.2.b]", (path ? &lpath_3 : NULL));
                return false;
            }
        }
        else if (strcmp(prop, "f") == 0)
        {
            // handle must f property
            must_count += 1;
            // $.'|'.2.f
            res = json_is_string(pval);
            if (! res)
            {
                if (rep) jm_report_add_entry(rep, "unexpected string [$.'|'.2.f]", (path ? &lpath_3 : NULL));
            }
            if (! res)
            {
                if (rep) jm_report_add_entry(rep, "invalid must property value [$.'|'.2.f]", (path ? &lpath_3 : NULL));
                return false;
            }
        }
        else
        {
            if (rep) jm_report_add_entry(rep, "no other prop expected [$.'|'.2]", (path ? &lpath_3 : NULL));
            return false;
        }
    }
    if (must_count != 3)
    {
        if (rep) jm_report_add_entry(rep, "missing must prop [$.'|'.2]", path);
        return false;
    }
    return true;
}

// object $.'|'.1
static bool _jm_obj_4(const json_t* val, Path* path, Report* rep)
{
    if (! json_is_object(val))
    {
        if (rep) jm_report_add_entry(rep, "not an object [$.'|'.1]", path);
        return false;
    }
    bool res;
    int64_t must_count = 0;
    const char *prop;
    json_t *pval;
    json_object_foreach((json_t *) val, prop, pval)
    {
        Path lpath_4 = (Path) { prop, 0, path, NULL };
        if (strcmp(prop, "a") == 0)
        {
            // handle must a property
            must_count += 1;
            // $.'|'.1.a
            res = json_is_integer(pval) && json_integer_value(pval) >= 0;
            if (! res)
            {
                if (rep) jm_report_add_entry(rep, "not a 0 strict int [$.'|'.1.a]", (path ? &lpath_4 : NULL));
            }
            if (! res)
            {
                if (rep) jm_report_add_entry(rep, "invalid must property value [$.'|'.1.a]", (path ? &lpath_4 : NULL));
                return false;
            }
        }
        else if (strcmp(prop, "b") == 0)
        {
            // handle must b property
            must_count += 1;
            // $.'|'.1.b
            res = json_is_integer(pval) && json_integer_value(pval) >= 1;
            if (! res)
            {
                if (rep) jm_report_add_entry(rep, "not a 1 strict int [$.'|'.1.b]", (path ? &lpath_4 : NULL));
            }
            if (! res)
            {
                if (rep) jm_report_add_entry(rep, "invalid must property value [$.'|'.1.b]", (path ? &lpath_4 : NULL));
                return false;
            }
        }
        else if (strcmp(prop, "e") == 0)
        {
            // handle must e property
            must_count += 1;
            // $.'|'.1.e
            res = json_is_string(pval);
            if (! res)
            {
                if (rep) jm_report_add_entry(rep, "unexpected string [$.'|'.1.e]", (path ? &lpath_4 : NULL));
            }
            if (! res)
            {
                if (rep) jm_report_add_entry(rep, "invalid must property value [$.'|'.1.e]", (path ? &lpath_4 : NULL));
                return false;
            }
        }
        else
        {
            if (rep) jm_report_add_entry(rep, "no other prop expected [$.'|'.1]", (path ? &lpath_4 : NULL));
            return false;
        }
    }
    if (must_count != 3)
    {
        if (rep) jm_report_add_entry(rep, "missing must prop [$.'|'.1]", path);
        return false;
    }
    return true;
}

// object $.'|'.0
static bool _jm_obj_5(const json_t* val, Path* path, Report* rep)
{
    if (! json_is_object(val))
    {
        if (rep) jm_report_add_entry(rep, "not an object [$.'|'.0]", path);
        return false;
    }
    bool res;
    int64_t must_count = 0;
    const char *prop;
    json_t *pval;
    json_object_foreach((json_t *) val, prop, pval)
    {
        Path lpath_5 = (Path) { prop, 0, path, NULL };
        if (strcmp(prop, "a") == 0)
        {
            // handle must a property
            must_count += 1;
            // $.'|'.0.a
            res = json_is_integer(pval) && json_integer_value(pval) >= 0;
            if (! res)
            {
                if (rep) jm_report_add_entry(rep, "not a 0 strict int [$.'|'.0.a]", (path ? &lpath_5 : NULL));
            }
            if (! res)
            {
                if (rep) jm_report_add_entry(rep, "invalid must property value [$.'|'.0.a]", (path ? &lpath_5 : NULL));
                return false;
            }
        }
        else if (strcmp(prop, "b") == 0)
        {
            // handle must b property
            must_count += 1;
            // $.'|'.0.b
            res = json_is_integer(pval) && json_integer_value(pval) >= 1;
            if (! res)
            {
                if (rep) jm_report_add_entry(rep, "not a 1 strict int [$.'|'.0.b]", (path ? &lpath_5 : NULL));
            }
            if (! res)
            {
                if (rep) jm_report_add_entry(rep, "invalid must property value [$.'|'.0.b]", (path ? &lpath_5 : NULL));
                return false;
            }
        }
        else if (strcmp(prop, "d") == 0)
        {
            // handle must d property
            must_count += 1;
            // $.'|'.0.d
            res = json_is_string(pval);
            if (! res)
            {
                if (rep) jm_report_add_entry(rep, "unexpected string [$.'|'.0.d]", (path ? &lpath_5 : NULL));
            }
            if (! res)
            {
                if (rep) jm_report_add_entry(rep, "invalid must property value [$.'|'.0.d]", (path ? &lpath_5 : NULL));
                return false;
            }
        }
        else
        {
            if (rep) jm_report_add_entry(rep, "no other prop expected [$.'|'.0]", (path ? &lpath_5 : NULL));
            return false;
        }
    }
    if (must_count != 3)
    {
        if (rep) jm_report_add_entry(rep, "missing must prop [$.'|'.0]", path);
        return false;
    }
    return true;
}

// check $ ($)
static bool json_model_1(const json_t* val, Path* path, Report* rep)
{
    bool res;
    // $
    res = json_is_object(val);
    if (! res)
    {
        if (rep) jm_report_add_entry(rep, "unexpected type at [$.'|']", path);
    }
    if (res)
    {
        // $.'|'.0
        res = _jm_obj_5(val, path, rep);
        if (! res)
        {
            if (rep) jm_report_add_entry(rep, "not an expected object at [$.'|'.0]", path);
        }
        if (! res)
        {
            // $.'|'.1
            res = _jm_obj_4(val, path, rep);
            if (! res)
            {
                if (rep) jm_report_add_entry(rep, "not an expected object at [$.'|'.1]", path);
            }
            if (! res)
            {
                // $.'|'.2
                res = _jm_obj_3(val, path, rep);
                if (! res)
                {
                    if (rep) jm_report_add_entry(rep, "not an expected object at [$.'|'.2]", path);
                }
                if (! res)
                {
                    // $.'|'.3
                    res = _jm_obj_2(val, path, rep);
                    if (! res)
                    {
                        if (rep) jm_report_add_entry(rep, "not an expected object at [$.'|'.3]", path);
                    }
                    if (! res)
                    {
                        // $.'|'.4
                        res = _jm_obj_1(val, path, rep);
                        if (! res)
                        {
                            if (rep) jm_report_add_entry(rep, "not an expected object at [$.'|'.4]", path);
                        }
                        if (! res)
                        {
                            // $.'|'.5
                            res = _jm_obj_0(val, path, rep);
                            if (! res)
                            {
                                if (rep) jm_report_add_entry(rep, "not an expected object at [$.'|'.5]", path);
                            }
                        }
                    }
                }
            }
        }
        if (! res)
        {
            if (rep) jm_report_add_entry(rep, "no model matched [$.'|']", path);
        }
    }
    return res;
}

check_fun_t check_model_map(const char *pname)
{
    return jm_search_propmap(pname, check_model_map_tab, 1);
}

static bool initialized = false;

char *CHECK_init(void)
{
    if (!initialized)
    {
        initialized = true;
        jm_version_string = JSON_MODEL_VERSION;
        check_model_map_tab[0] = (propmap_t) { "", json_model_1 };
        jm_sort_propmap(check_model_map_tab, 1);
    }
    return NULL;
}

void CHECK_free(void)
{
    if (initialized)
    {
        initialized = false;

        // cleanup code
    }
}


/*
 * # API
 *
 * ```c
 * bool error;
 * char *message;
 *
 * bool valid = check(json_value, "model-name", &error, &message);
 *
 * if (message)
 *    free(message), message = NULL;
 * ```
 *
 * - `error` is set of there was an error during initialization or if the model does not exists.
 * - `message` is provided if a non NULL pointer is passed, and must be freed by the caller.
 */
bool
CHECK(const json_t *val, const char *name, bool *error, char **reasons)
{
    return jm_generic_entry(CHECK_init, CHECK_fun, val, name, error, reasons);
}

File: tests/mv-00/dst_01.pl
#! /usr/bin/perl
use strict;
use warnings;
no warnings "uninitialized";
use JSON::JsonModel;
use constant JMC_VERSION => '2';


sub _jm_obj_0($$$);
sub _jm_obj_1($$$);
sub _jm_obj_2($$$);
sub _jm_obj_3($$$);
sub _jm_obj_4($$$);
sub _jm_obj_5($$$);
sub json_model_1($$$);
my %check_model_map;

sub _jm_obj_0($$$)
{
    my ($val, $path, $rep) = @_;
    if (jm_obj_size($val) != 3)
    {
        return 0;
    }
    my $pval;
    if (! exists $$val{'a'})
    {
        return 0;
    }
    $pval = $$val{'a'};
    my $res = jm_is_integer($pval) && $pval >= 0;
    if (! $res)
    {
        return 0;
    }
    if (! exists $$val{'c'})
    {
        return 0;
    }
    $pval = $$val{'c'};
    $res = jm_is_integer($pval) && $pval >= 1;
    if (! $res)
    {
        return 0;
    }
    if (! exists $$val{'f'})
    {
        return 0;
    }
    $pval = $$val{'f'};
    return jm_is_string($pval);
}

sub _jm_obj_1($$$)
{
    my ($val, $path, $rep) = @_;
    if (jm_obj_size($val) != 3)
    {
        return 0;
    }
    my $pval;
    if (! exists $$val{'a'})
    {
        return 0;
    }
    $pval = $$val{'a'};
    my $res = jm_is_integer($pval) && $pval >= 0;
    if (! $res)
    {
        return 0;
    }
    if (! exists $$val{'c'})
    {
        return 0;
    }
    $pval = $$val{'c'};
    $res = jm_is_integer($pval) && $pval >= 1;
    if (! $res)
    {
        return 0;
    }
    if (! exists $$val{'e'})
    {
        return 0;
    }
    $pval = $$val{'e'};
    return jm_is_string($pval);
}

sub _jm_obj_2($$$)
{
    my ($val, $path, $rep) = @_;
    if (jm_obj_size($val) != 3)
    {
        return 0;
    }
    my $pval;
    if (! exists $$val{'a'})
    {
        return 0;
    }
    $pval = $$val{'a'};
    my $res = jm_is_integer($pval) && $pval >= 0;
    if (! $res)
    {
        return 0;
    }
    if (! exists $$val{'c'})
    {
        return 0;
    }
    $pval = $$val{'c'};
    $res = jm_is_integer($pval) && $pval >= 1;
    if (! $res)
    {
        return 0;
    }
    if (! exists $$val{'d'})
    {
        return 0;
    }
    $pval = $$val{'d'};
    return jm_is_string($pval);
}

sub _jm_obj_3($$$)
{
    my ($val, $path, $rep) = @_;
    if (jm_obj_size($val) != 3)
    {
        return 0;
    }
    my $pval;
    if (! exists $$val{'a'})
    {
        return 0;
    }
    $pval = $$val{'a'};
    my $res = jm_is_integer($pval) && $pval >= 0;
    if (! $res)
    {
        return 0;
    }
    if (! exists $$val{'b'})
    {
        return 0;
    }
    $pval = $$val{'b'};
    $res = jm_is_integer($pval) && $pval >= 1;
    if (! $res)
    {
        return 0;
    }
    if (! exists $$val{'f'})
    {
        return 0;
    }
    $pval = $$val{'f'};
    return jm_is_string($pval);
}

sub _jm_obj_4($$$)
{
    my ($val, $path, $rep) = @_;
    if (jm_obj_size($val) != 3)
    {
        return 0;
    }
    my $pval;
    if (! exists $$val{'a'})
    {
        return 0;
    }
    $pval = $$val{'a'};
    my $res = jm_is_integer($pval) && $pval >= 0;
    if (! $res)
    {
        return 0;
    }
    if (! exists $$val{'b'})
    {
        return 0;
    }
    $pval = $$val{'b'};
    $res = jm_is_integer($pval) && $pval >= 1;
    if (! $res)
    {
        return 0;
    }
    if (! exists $$val{'e'})
    {
        return 0;
    }
    $pval = $$val{'e'};
    return jm_is_string($pval);
}

sub _jm_obj_5($$$)
{
    my ($val, $path, $rep) = @_;
    if (jm_obj_size($val) != 3)
    {
        return 0;
    }
    my $pval;
    if (! exists $$val{'a'})
    {
        return 0;
    }
    $pval = $$val{'a'};
    my $res = jm_is_integer($pval) && $pval >= 0;
    if (! $res)
    {
        return 0;
    }
    if (! exists $$val{'b'})
    {
        return 0;
    }
    $pval = $$val{'b'};
    $res = jm_is_integer($pval) && $pval >= 1;
    if (! $res)
    {
        return 0;
    }
    if (! exists $$val{'d'})
    {
        return 0;
    }
    $pval = $$val{'d'};
    return jm_is_string($pval);
}

sub json_model_1($$$)
{
    my ($val, $path, $rep) = @_;
    return jm_is_object($val) && (exists $$val{'d'} && _jm_obj_5($val, undef, undef) || exists $$val{'e'} && _jm_obj_4($val, undef, undef) || exists $$val{'f'} && _jm_obj_3($val, undef, undef) || exists $$val{'d'} && _jm_obj_2($val, undef, undef) || exists $$val{'e'} && _jm_obj_1($val, undef, undef) || exists $$val{'f'} && _jm_obj_0($val, undef, undef));
}


# initialization of global variables

our $initialized = 0;

sub check_model_init()
{
    if (!$initialized)
    {
        $initialized = 1;
        %check_model_map = (
            '' => \&json_model_1,
        );
    }
}

sub check_model_free()
{
    if ($initialized)
    {
        $initialized = 0;
        %check_model_map = ();
    }
}

sub check_model_mapper($)
{
    my ($name) = @_;
    die "unexpected model name \"$name\"" unless exists $check_model_map{$name};
    return $check_model_map{$name};
}

sub check_model($$$)
{
    my ($json, $name, $rep) = @_;
    die "unexpected model name \"$name\"" unless exists $check_model_map{$name};
    my $path = defined $rep ? [] : undef;
    return $check_model_map{$name}($json, $path, $rep);
}

#
# SCRIPT EXECUTION
#
check_model_init();
jm_main(\&check_model, \%check_model_map, JMC_VERSION);
check_model_free();
